src/python/pants/engine/scheduler.py

```diff
@@ -268,63 +226,34 @@ def graph_len(self):
+        self._native.lib.execution_add_root_select(
+            self._scheduler, execution_request, params, product
         )
-        self._raise_or_return(res)
 
     @property
     def visualize_to_dir(self):
         return self._visualize_to_dir
 
     def _metrics(self, session):
-        return self._from_value(self._native.lib.scheduler_metrics(self._scheduler, session))
+        return self._native.lib.scheduler_metrics(self._scheduler, session)
 
     def poll_workunits(self, session) -> PolledWorkunits:
-        result: Tuple[Tuple[Workunit], Tuple[Workunit]] = self._from_value(
-            self._native.lib.poll_session_workunits(self._scheduler, session)
+        result: Tuple[Tuple[Workunit], Tuple[Workunit]] = self._native.lib.poll_session_workunits(
+            self._scheduler, session
         )
         return {"started": result[0], "completed": result[1]}
 
     def _run_and_return_roots(self, session, execution_request):
         raw_roots = self._native.lib.scheduler_execute(self._scheduler, session, execution_request)
-        if raw_roots == self._native.ffi.NULL:
-            raise KeyboardInterrupt
 
-        remaining_runtime_exceptions_to_capture = list(
-            self._native.consume_cffi_extern_method_runtime_exceptions()
-        )
-        try:
-            roots = []
-            for raw_root in self._native.unpack(raw_roots.nodes_ptr, raw_roots.nodes_len):
-                # Check if there were any uncaught exceptions within rules that were executed.
-                remaining_runtime_exceptions_to_capture.extend(
-                    self._native.consume_cffi_extern_method_runtime_exceptions()
-                )
-
-                if raw_root.is_throw:
-                    state = Throw(self._from_value(raw_root.handle))
-                elif raw_root.handle == self._native.ffi.NULL:
-                    # NB: We expect all NULL handles to correspond to uncaught exceptions which are collected
-                    # in `self._native._peek_cffi_extern_method_runtime_exceptions()`!
-                    if not remaining_runtime_exceptions_to_capture:
-                        raise ExecutionError(
-                            "Internal logic error in scheduler: expected more elements in "
-                            "`self._native._peek_cffi_extern_method_runtime_exceptions()`."
-                        )
-                    matching_runtime_exception = remaining_runtime_exceptions_to_capture.pop(0)
-                    state = Throw(matching_runtime_exception)
-                else:
-                    state = Return(self._from_value(raw_root.handle))
-                roots.append(state)
-        finally:
-            self._native.lib.nodes_destroy(raw_roots)
+        roots = []
+        for raw_root in raw_roots:
+            if raw_root.is_throw():
+                state = Throw(raw_root.handle())
+            else:
+                state = Return(raw_root.handle())
+            roots.append(state)
```

Comment:
This whole thing can be a list comprehension now thanks to how much simpler you've made it :)

```python
roots = [
  Throw(raw_root.handle()) if raw_root.is_throw() else Return(raw_root.handle())
  for raw_root in raw_roots
]
```